File: backend/app/schemas/prompts.py
"""Prompt generation Pydantic schemas.

Faithfully replicates the n8n "Raw to Promotable" workflow's
data structures for prompt generation.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class WorkflowAnalysisData(BaseModel):
    """11-field analysis schema from the n8n workflow's Gemini Vision node."""

    productIdentity: str = Field(default="", alias="Product Identity")
    materialProperties: str = Field(default="", alias="Material Properties")
    scaleAndProportion: str = Field(default="", alias="Scale and Proportion")
    designStyle: str = Field(default="", alias="Design Style")
    visualCraftsmanship: str = Field(default="", alias="Visual Craftsmanship")
    targetDemographic: str = Field(default="", alias="Target Demographic")
    psychographics: str = Field(default="", alias="Psychographics")
    functionalUtility: str = Field(default="", alias="Functional Utility")
    lifestyleBranding: str = Field(default="", alias="Lifestyle Branding")
    indianFestiveContext: str = Field(default="", alias="Indian Festive Context")
    marketReadiness: str = Field(default="", alias="Market Readiness")

    class Config:
        populate_by_name = True
        extra = "ignore"


class PromptGenerateRequest(BaseModel):
    """Request to generate promotional prompts for an image."""

    image_id: str = Field(..., description="ID of the uploaded image")
    description: Optional[str] = Field(None, description="Optional consumer description")


class PromptGenerateResponse(BaseModel):
    """Response containing generated prompts for all 8 categories."""

    success: bool = Field(default=True)
    workflowAnalysis: Optional[WorkflowAnalysisData] = Field(None, alias="workflow_analysis")
    prompts: Optional[Dict[str, str]] = Field(None, description="Map of category -> generated prompt")
    generationTimeMs: float = Field(default=0.0, alias="generation_time_ms")
    imageId: str = Field(default="", alias="image_id")
    error: Optional[str] = Field(None)

    class Config:
        populate_by_name = True


class PromptProcessingResponse(BaseModel):
    """Response returned while prompt generation is processing."""

    status: str = Field("processing")
    imageId: str = Field(..., alias="image_id")
    requestId: str = Field(..., alias="request_id")

    class Config:
        populate_by_name = True

File: frontend/src/types/prompts.ts
// ============================================================
// prompts.ts — Prompt Generation Type Definitions
// MORAA GemVision
// ============================================================
// Faithfully replicates the n8n "Raw to Promotable" workflow's
// Gemini Vision Analysis output schema and all 8 prompt categories.
// ============================================================

// ─── Workflow Analysis Schema ─────────────────────────────
// This is the exact 11-field JSON schema the n8n workflow's
// Gemini Vision Analysis node extracts from the image.

export interface WorkflowAnalysisData {
  productIdentity: string;
  materialProperties: string;
  scaleAndProportion: string;
  designStyle: string;
  visualCraftsmanship: string;
  targetDemographic: string;
  psychographics: string;
  functionalUtility: string;
  lifestyleBranding: string;
  indianFestiveContext: string;
  marketReadiness: string;
}

// ─── Prompt Categories ────────────────────────────────────
// Exactly matching the 8 prompt generation nodes in the workflow.

export const PROMPT_CATEGORIES = [
  "professionalShot",
  "useCaseShot",
  "ingredientStory",
  "festive",
  "transformation",
  "scaleReference",
  "complementaryShot",
  "ugcStyle",
] as const;

export type PromptCategory = (typeof PROMPT_CATEGORIES)[number];

export const PROMPT_CATEGORY_LABELS: Record<PromptCategory, string> = {
  professionalShot: "Professional Shot",
  useCaseShot: "Use Case Shot",
  ingredientStory: "Craft Story",
  festive: "Festive",
  transformation: "Transformation",
  scaleReference: "Scale Reference",
  complementaryShot: "Complementary Shot",
  ugcStyle: "UGC Style",
};

export const PROMPT_CATEGORY_DESCRIPTIONS: Record<PromptCategory, string> = {
  professionalShot:
    "Premium studio photoshoot — controlled lighting, refined props, editorial composition",
  useCaseShot:
    "Product in authentic use — a real person interacting naturally with the product",
  ingredientStory:
    "Making-of / provenance — finished product + its raw material or craft process",
  festive:
    "Indian festival campaign — product within a festive cultural moment",
  transformation:
    "Before/after emotional shift — split-frame showing product's transformative effect",
  scaleReference:
    "Scale accuracy — product next to a familiar reference object for true size",
  complementaryShot:
    "Lifestyle pairing — product alongside its most natural companion item",
  ugcStyle:
    "User-generated content — casual smartphone shot, authentic everyday setting",
};

// ─── API Types ─────────────────────────────────────────────

export interface PromptGenerateRequest {
  imageBase64: string;
  mimeType: string;
  description?: string;
}

export interface PromptGenerateResponse {
  success: boolean;
  data: {
    workflowAnalysis: WorkflowAnalysisData;
    prompts: Record<PromptCategory, string>;
    generationTimeMs: number;
  } | null;
  error?: {
    code: string;
    message: string;
  };
}

export interface PromptGenerationState {
  status: "idle" | "analyzing" | "generating" | "completed" | "error";
  errorMessage: string | null;
  workflowAnalysis: WorkflowAnalysisData | null;
  prompts: Record<PromptCategory, string> | null;
  generationTimeMs: number;
}
